Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs

```diff
@@ -174,118 +171,161 @@ public IConnection Acquire(AccessMode mode)
             return Acquire();
         }
 
-        public Task<IConnection> AcquireAsync(AccessMode mode)
+        public IPooledConnection Acquire()
         {
-            return AcquireAsync();
+            using (var timeOutTokenSource = new CancellationTokenSource(_connAcquisitionTimeout))
+            {
+                return Acquire(timeOutTokenSource.Token);
+            }
         }
 
-        public IPooledConnection Acquire()
+        private IPooledConnection Acquire(CancellationToken cancellationToken)
         {
             return TryExecute(() =>
             {
-                if (_disposeCalled)
-                {
-                    ThrowObjectDisposedException();
-                }
-                IPooledConnection connection;
+                IPooledConnection connection = null;
 
-                if (!_availableConnections.TryTake(out connection))
+                try
                 {
-                    var connAcquisitionTimer = new Stopwatch();
-                    connAcquisitionTimer.Start();
-                    do
+                    while (true)
                     {
-                        if (!IsConnectionPoolFull())
+                        if (_disposeCalled)
                         {
-                            connection = CreateNewPooledConnection();
-                            break;
+                            ThrowObjectDisposedException();
+                        }
+
+                        if (!_availableConnections.TryTake(out connection))
+                        {
+                            do
+                            {
+                                if (!IsConnectionPoolFull())
+                                {
+                                    connection = CreateNewPooledConnection();
```

Comment:
Unrelated to this PR: given a pool with `maxConnectionPoolSize - 1` acquired and `0` idle connections, can two threads that call `Acquire()` concurrently make pool expand to `maxConnectionPoolSize + 1` active connections? They will both try to create concurrently and attempts are probably not synchronized.

Reply:
Except for correctness, I am not fan to strictly not exceed maxConnPoolSize. As it not only slows the code down also bring more complete logic.

Reply:
Suggesting to have the fix in another PR. Right now the scope of this PR is good enough and the problem is not critical to block 1.5.